### PR TITLE
AIP-51 - Misc. Compatibility Checks

### LINKED ISSUE #27930: AIP-51 - Misc. Compatibility Checks
### Overview

Coupling around other compatibility checks, similar to coupling 1 and 2, but which don't fall into either category.

### Examples
- **3a**) Some Executors support pickling, some don’t, there is logic in both backfill and scheduler jobs that checks before submitting tasks to the Executor queue. Applies to DaskExecutor, LocalExecutor and SequentialExecutor as of now: https://github.com/apache/airflow/blob/26f94c5370587f73ebd935cecf208c6a36bdf9b6/airflow/jobs/backfill_job.py#L825-L829 https://github.com/apache/airflow/blob/26f94c5370587f73ebd935cecf208c6a36bdf9b6/airflow/jobs/scheduler_job.py#L703-L704
- **3b**) Sentry has a celery integration, if the Executor is Celery, the Sentry integration is imported and configured: https://github.com/apache/airflow/blob/27e2101f6ee5567b2843cbccf1dca0b0e7c96186/airflow/sentry.py#L87
  - Sentry at the moment only supports Celery, so this particular coupling is low priority, but should still be fixed.

### Proposal
A static method or attribute on the Executor class which can be checked by core code.

There is a precedent already set with the `supports_ad_hoc_ti_run` attribute, see:
https://github.com/apache/airflow/blob/fb741fd87254e235f99d7d67e558dafad601f253/airflow/executors/kubernetes_executor.py#L435 https://github.com/apache/airflow/blob/26f94c5370587f73ebd935cecf208c6a36bdf9b6/airflow/www/views.py#L1735-L1737

### PR DESCRIPTION
**Pull request by rkarish**
Closes: apache#27930

### CODE REVIEW

File: airflow/executors/celery_kubernetes_executor.py
Comment (o-nikolas) [inline airflow/executors/celery_kubernetes_executor.py]: You'll also need to set a value for `supports_sentry` in this class as well since it does not implement the `BaseExecutor` interface (see the discussion starting [from here](https://github.com/apache/airflow/issues/28276#issuecomment-1344899475) for more context). You'll need to ensure both new fields are set on all the executors which don't inherit from `BaseExecutor` (`LocalKubernetesExecutor` as well) and add the appropriate unit tests for those.

File: airflow/executors/executor_loader.py
Comment (o-nikolas) [inline airflow/executors/executor_loader.py:172]: I wonder if we should actually keep this in for backwards compatibility (with a comment describing that it is deprecated) :thinking: i.e. someone in the community could be depending on this list existing. I wonder what others think about this.

File: airflow/executors/base_executor.py
Comment (o-nikolas) [inline airflow/executors/base_executor.py]: I think we want the default to be True, no? The previous code assumed all executors were picklable unless it was present in the list of `UNPICKLEABLE_EXECUTORS`. So any executor that goes through that code should be assumed to be picklable unless it says otherwise. WDYT?

File: airflow/executors/local_kubernetes_executor.py
Comment (o-nikolas) [inline airflow/executors/local_kubernetes_executor.py]: Do we actually want this to be True? The `LocalExecutor` is not picklable and so far we've gone with True for these only if all sub executors are True. WDYT?
There is some context in a PR [comment here](https://github.com/apache/airflow/pull/28288/files#r1045147675) (from @pierrejeambrun's PR for `is_local`)

File: airflow/sentry.py
Comment (o-nikolas) [inline airflow/sentry.py]: You can use the helper method I added for this `ExecutorLoader.import_default_executor_cls()`

**Review by o-nikolas — COMMENTED** [review]
Thanks for the PR! I've left some comments and requests :)

File: airflow/executors/base_executor.py
Comment (rkarish) [reply airflow/executors/base_executor.py]: I agree that this makes more sense. For some reason I went with the opposite 🙃

File: airflow/executors/executor_loader.py
Comment (rkarish) [reply airflow/executors/executor_loader.py:172]: I searched the repository for references and it looked like it was safe to delete, but I didn't think about other users in the community being dependent on this. I think that adding a comment here and waiting to clean this up at a later time is a safer approach.

**Comment by rkarish** [comment]
Thanks for the feedback Niko! I will clean this up and let you know when I am finished.

**Comment by rkarish** [comment]
Hey Niko, I've submitted some more changes based on the feedback and requests. I reworked the pickling support attribute, switched to using your get executor class helper method, and implemented tests for pickling and sentry support for all of the executors. CI pipeline failed so I rebased the branch off main which fixed a failure in integration tests, but it's still failing static checks. Pre-commit was passing for me locally before rebasing, seems like something changed here.

Also, I didn't mean to remove all of the other reviewers, I clicked on request re-review and it did it automatically.

**Comment by o-nikolas** [comment]
> Hey Niko, I've submitted some more changes based on the feedback and requests. I reworked the pickling support attribute, switched to using your get executor class helper method, and implemented tests for pickling and sentry support for all of the executors. CI pipeline failed so I rebased the branch off main which fixed a failure in integration tests, but it's still failing static checks. Pre-commit was passing for me locally before rebasing, seems like something changed here. When I ran it locally after rebase it was changing several files so I didn't commit those changes.
> 
> Also, I didn't mean to remove all of the other reviewers, I clicked on request re-review and it did it automatically.

Thanks @rkarish! Sorry for the delayed response but I have been on year end/holiday vacation. I'm just trying to get caught up now. I'll have a look at this PR very soon :pray:

**Comment by rkarish** [comment]
You're welcome and Happy New Year! I figured you were on vacation, no rush on the review!

**Review by o-nikolas — APPROVED** [review]
The changes look great! Perhaps @pierrejeambrun can be the second reviewer/approver, he's been very involved in the executor decoupling work.

**Comment by rkarish** [comment]
That would be great to get another set of eyes on this. Congratulations on becoming a Committer by the way! Would love for you to do the honors on this one 😄 .

File: airflow/executors/executor_loader.py
Comment (pierrejeambrun) [reply airflow/executors/executor_loader.py:172]: Maybe add a `# TODO: Remove in Airflow 3.0`. We have a couple of those in the repo and will help to not miss this spot when releasing the next major.

**Review by pierrejeambrun — COMMENTED** [review]
Just one suggestion, otherwise LGTM :)

**Review by pierrejeambrun — APPROVED** [review]

**Comment by o-nikolas** [comment]
> That would be great to get another set of eyes on this. Congratulations on becoming a Committer by the way! Would love for you to do the honors on this one :smile:.

Thanks @rkarish! And thanks to you for helping on AIP-51 :grinning:

**Pull request merged**
